vcd/resource_vcd_edgegateway.go

```diff
@@ -0,0 +1,232 @@
+package vcd
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+)
+
+func resourceVcdEdgeGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVcdEdgeGatewayCreate,
+		Delete: resourceVcdEdgeGatewayDelete,
+		Read:   resourceVcdEdgeGatewayRead,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"org": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"vdc": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+			"advanced": &schema.Schema{
+				Type:        schema.TypeBool,
+				Required:    true,
+				ForceNew:    true,
+				Description: "True if the gateway uses advanced networking. (Set by default in vCD 9.7+)",
+			},
+			"gateway_configuration": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Configuration of the vShield edge VM for this gateway. One of: compact, full",
+			},
+			"ha_enabled": &schema.Schema{
+				Type:        schema.TypeBool,
+				Optional:    true,
+				Default:     false,
+				ForceNew:    true,
+				Description: "Enable high availability on this edge gateway",
+			},
+			"external_networks": &schema.Schema{
+				Type:        schema.TypeList,
+				Required:    true,
+				ForceNew:    true,
+				Description: "A list of external networks to be used by the edge gateway",
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"default_gateway": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Default:     "",
+				ForceNew:    true,
+				Description: "External network to be used as default gateway. Its name must be included in 'external_networks'. An empty value will skip the default gateway",
+			},
+			"distributed_routing": &schema.Schema{
```

Comment:
Something is wrong here (at least on vCD 9.5).
With the configuration (copied in the bottom) the resource always gets recreated (because `refresh` sets `distributed_routing` to `false` even though it was set to `true` in the config. ).

Replication:
* Use the `hcl` config in the bottom.
* `terraform apply` will create it
* `terraform apply` once again and you will see it wants to replace the resource because "`~ distributed_routing   = false -> true # forces replacement`"

P.S. maybe it's worth to include this argument in the test as well.

```
terraform apply
vcd_edgegateway.egw: Refreshing state... [id=urn:vcloud:gateway:d011190e-d1e1-45f9-b45d-b1803ce90023]

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
-/+ destroy and then create replacement

Terraform will perform the following actions:

  # vcd_edgegateway.egw must be replaced
-/+ resource "vcd_edgegateway" "egw" {
        advanced              = true
        default_gateway       = "my-ext-network"
        description           = "new edge gateway"
      ~ distributed_routing   = false -> true # forces replacement
        external_networks     = [
            "my-ext-network",
        ]
        gateway_configuration = "full"
        ha_enabled            = false
      ~ id                    = "urn:vcloud:gateway:d011190e-d1e1-45f9-b45d-b1803ce90023" -> (known after apply)
        name                  = "test-edge"
        org                   = "my-org"
        vdc                   = "my-org-vdc"
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```


```hcl
resource "vcd_edgegateway" "egw" {

  org = "my-org"
  vdc = "my-org-vdc"

  name              = "test-edge"
  description       = "new edge gateway"
  gateway_configuration    = "full"
  default_gateway   = "my-ext-network"
  external_networks = [ "my-ext-network" ]
  advanced = true
  ha_enabled = false
  distributed_routing = true
}
```

Reply:
Based on testing the API does accept (and set) the setting, but upon GET query it omits the DistributedRouting field therefore struct field defaults to `false` after unmarshaling.
This has already been a case for us and then it was proven that API v27.0 does not return field, while API v31.0 does return the field.
In the light of this we can currently comment out line `err = d.Set("distributed_routing", egw.EdgeGateway.Configuration.DistributedRoutingEnabled)` (https://github.com/terraform-providers/terraform-provider-vcd/pull/263/files#diff-d38bb674f544446ef65cd837d6dfc207R188) with adding a //TODO for when we switch to higher API version.

That way we will rely on the fact, that whatever the user has set in HCL configuration was actually set correctly. The caveat is that if the user changes this setting via UI, `terraform refresh` will not pick the difference, but I'm afraid we're locked to this with API v27.0

Reply:
Addressed as suggested.

Reply:
Thanks. At least for me it works as expected now. I can provision with distributed_routing enabled and disabled as well as subsequent `terraform apply` does not ask for in place updates anymore.